File: src/horizon_mcp/tools/assist/computation.py
"""Computation and datasource flow simulation tools.

2 tools for testing computation rule expressions and datasource flow
pipelines against sample data before deploying them in production profiles.

Knowledge resources:
    - horizon://knowledge/computation-and-data-flow
"""

from __future__ import annotations

import json
import logging
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from mcp.server.fastmcp import FastMCP

logger = logging.getLogger("horizon_mcp.tools.assist.computation")


def register_computation_tools(mcp: FastMCP) -> None:
    """Register computation simulation tools on *mcp*."""

    from horizon_mcp.client.state import get_client

    @mcp.tool()
    async def simulate_computation_rule(
        rule: str,
        dictionary: dict,
        mode: str = "computation_rule",
    ) -> str:
        """Test a computation rule or template string against a dictionary.

        See horizon://knowledge/computation-and-data-flow for syntax and
        available functions.

        Safety tier: read-only

        Horizon has two expression types:

        **computation_rule** (default): Full expression language with functions.
        Used in profile certificate templates to compute field values.
        Examples:
            - ``Upper({{cn}})`` — uppercase the cn dictionary value
            - ``DomainDNS({{fqdn}})`` — extract parent domain
            - ``Concat({{a}}, "-", {{b}})`` — concatenate values
            - ``{{owner}}`` — simple dictionary lookup
            - ``OrElse({{prefix}}, "default")`` — fallback chain

        **template_string**: Simple text interpolation with ``{{key}}`` placeholders.
        Used in email templates, webhook URLs, notification bodies.
        Examples:
            - ``Hello {{name}}, your cert expires on {{certificate.not_after}}``
            - ``https://api.example.com/v1/{{certificate.serial}}``

        Args:
            rule: The expression to evaluate. For computation rules, use
                  function calls with ``{{key}}`` for dictionary lookups.
                  For template strings, use free text with ``{{key}}`` placeholders.
            dictionary: Key-value pairs available as variables during evaluation.
                        All values must be strings.
            mode: Expression type — "computation_rule" (default) or "template_string".

        Returns:
            JSON with the computed result including computedValueSingle,
            optionally computedValueMulti (for computation rules), and
            the merged dictionary used during evaluation.
        """
        valid_modes = {"computation_rule", "template_string"}
        if mode not in valid_modes:
            return json.dumps({
                "error": True,
                "content": f"Invalid mode '{mode}'. Must be one of: {', '.join(sorted(valid_modes))}.",
            })

        client = get_client()
        key = "computationRule" if mode == "computation_rule" else "templateString"
        result = await client.post(
            "/api/v1/templatestring/playground",
            json={key: rule, "dictionary": dictionary},
        )
        return json.dumps(result)

    @mcp.tool()
    async def simulate_datasource_flow(
        flow: list[dict],
        context: dict[str, Any] | None = None,
    ) -> str:
        """Test a datasource flow pipeline against an optional context.

        Safety tier: read-only

        Executes a datasource flow chain in test mode and returns the
        enriched dictionary. Each flow entry specifies a datasource name,
        input mappings, and an optional stop-on-success flag.

        Args:
            flow: Ordered list of flow entries. Each entry is a dict with
                  keys ``datasource`` (str), ``inputs`` (dict[str, str]),
                  and optionally ``stopOnSuccess`` (bool).
            context: Optional initial context dictionary providing seed
                     values for the flow evaluation.

        Returns:
            JSON with the enriched dictionary produced by the flow.
        """
        client = get_client()
        body: dict[str, Any] = {"flow": flow}
        if context is not None:
            body["context"] = context
        result = await client.post("/api/v1/datasources/flow/test", json=body)
        return json.dumps(result)

File: src/tools/assist/computation.ts
import type { McpServer } from '@modelcontextprotocol/server';
import { z } from 'zod';

import type { HorizonClient } from '../../client/http.js';
import { registerTool } from '../register.js';

function toMapEntries(
  values: Record<string, unknown> | undefined,
): Array<{ key: string; value: string }> | undefined {
  if (values === undefined) return undefined;
  return Object.entries(values)
    .map(([key, value]) => ({ key, value: String(value) }))
    .sort((a, b) => a.key.localeCompare(b.key));
}

function toDatasourceFlow(
  flow: Array<{
    datasource: string;
    inputs: Record<string, string>;
    stopOnSuccess: boolean;
  }>,
): Array<{
  ds: string;
  inputs?: Array<{ key: string; value: string }>;
  stopOnSuccess: boolean;
}> {
  return flow.map(({ datasource, inputs, stopOnSuccess }) => {
    const normalizedInputs = Object.entries(inputs)
      .map(([key, value]) => ({ key, value }))
      .sort((a, b) => a.key.localeCompare(b.key));
    return {
      ds: datasource,
      inputs: normalizedInputs.length > 0 ? normalizedInputs : undefined,
      stopOnSuccess,
    };
  });
}

export function registerComputationTools(
  server: McpServer,
  client: HorizonClient,
): void {
  registerTool(
    server,
    'simulate_computation_rule',
    {
      description:
        'Test a computation rule or template string against a dictionary.\n\n' +
        'MANDATORY: Before writing ANY computation rule, you MUST read the ' +
        'knowledge resource horizon://knowledge/computation-and-data-flow. ' +
        'It contains the COMPLETE list of available functions, the exact syntax, ' +
        'and real-world PKI examples. DO NOT invent functions or syntax - only ' +
        'use what is documented in that resource.\n\n' +
        'Available functions (exhaustive list - no others exist):\n' +
        '  String: Upper, Lower, Trim, Substr, Concat, Extract, Replace, OrElse\n' +
        '  List: Filter, Slice, Sort, Split, Unique\n' +
        '  Parsing: ShortenDNS, DomainDNS, EmailUser, EmailDomain, SamAccountNameUser, SamAccountNameDomain\n' +
        '  Date: DateTimeFormat\n' +
        '  Access: Get, First, Last, Join, Match\n' +
        '  Encoding: URLEncode, URLDecode, EscapeJson, JsonArray, DerAsBase64, Base64, Raw\n' +
        '  Special: NULL, NOW\n\n' +
        'Syntax rules:\n' +
        '  - Dictionary lookups: {{key}} for single, [[key]] for multi\n' +
        '  - Functions wrap lookups: Upper({{cn}}), NOT {{Upper(cn)}}\n' +
        '  - Concat on arrays merges them: Concat([[a]], [[b]]) -> combined list\n' +
        '  - Concat with null returns null: use OrElse({{key}}, "") to guard\n' +
        '  - ShortenDNS extracts hostname: ShortenDNS({{fqdn}}) -> first DNS label\n' +
        '  - DomainDNS extracts domain: DomainDNS({{fqdn}}) -> parent domain\n' +
        '  - Sort alphabetically sorts a list\n' +
        '  - Unique deduplicates a list\n\n' +
        'Two expression modes:\n\n' +
        '  computation_rule (default): Full expression language with functions.\n' +
        '    Upper({{cn}}) - DomainDNS({{fqdn}}) - Sort(Unique([[sans]]))\n\n' +
        '  template_string: Text interpolation with embedded {{ }} blocks.\n' +
        '    Hello {{name}}, cert expires {{certificate.not_after}}',
      inputSchema: z.object({
        rule: z
          .string()
          .describe(
            'The expression to evaluate. For computation rules, use function calls with {{key}} for dictionary lookups.',
          ),
        dictionary: z
          .record(z.string(), z.string())
          .describe(
            'Key-value pairs available as variables during evaluation.',
          ),
        mode: z
          .enum(['computation_rule', 'template_string'])
          .default('computation_rule')
          .describe(
            'Expression type - "computation_rule" or "template_string".',
          ),
      }),
    },
    async ({ rule, dictionary, mode }) => {
      const key =
        mode === 'computation_rule' ? 'computationRule' : 'templateString';
      const result = await client.post('/api/v1/templatestring/playground', {
        [key]: rule,
        dictionary,
      });
      return {
        content: [{ type: 'text' as const, text: JSON.stringify(result) }],
      };
    },
  );

  registerTool(
    server,
    'simulate_datasource_flow',
    {
      description:
        'Test a datasource flow pipeline against an optional context.\n\n' +
        'Executes a datasource flow chain in test mode and returns the ' +
        'enriched dictionary. Each flow entry specifies a datasource name, ' +
        'input mappings, and an optional stop-on-success flag. The MCP ' +
        "accepts a small-model-friendly shape and translates it to Horizon's " +
        'raw `dsFlow` request body.',
      inputSchema: z.object({
        flow: z
          .array(
            z.object({
              datasource: z.string(),
              inputs: z.record(z.string(), z.string()).default({}),
              stopOnSuccess: z.boolean().default(false),
            }),
          )
          .describe('Ordered list of flow entries.'),
        context: z
          .record(z.string(), z.unknown())
          .optional()
          .describe('Optional initial context dictionary.'),
      }),
    },
    async ({ flow, context }) => {
      const body: Record<string, unknown> = { dsFlow: toDatasourceFlow(flow) };
      const normalizedContext = toMapEntries(context);
      if (normalizedContext !== undefined) body['context'] = normalizedContext;
      const result = await client.post('/api/v1/datasource/flows', body);
      return {
        content: [{ type: 'text' as const, text: JSON.stringify(result) }],
      };
    },
  );
}
